service/src/main/java/bio/terra/externalcreds/services/JwtUtils.java

```diff
@@ -0,0 +1,147 @@
+import org.springframework.security.oauth2.jwt.JwtDecoders;
+import org.springframework.security.oauth2.jwt.JwtException;
+import org.springframework.stereotype.Service;
+
+@Service
+public class JwtUtils {
+  public static final String PASSPORT_JWT_V11_CLAIM = "passport_jwt_v11";
+  public static final String GA4GH_PASSPORT_V1_CLAIM = "ga4gh_passport_v1";
+  public static final String GA4GH_VISA_V1_CLAIM = "ga4gh_visa_v1";
+  public static final String VISA_TYPE_CLAIM = "type";
+  public static final String JKU_HEADER = "jku";
+
+  private final ExternalCredsConfig externalCredsConfig;
+
+  public JwtUtils(ExternalCredsConfig externalCredsConfig) {
+    this.externalCredsConfig = externalCredsConfig;
+  }
+
+  public LinkedAccountWithPassportAndVisas enrichAccountWithPassportAndVisas(
+      LinkedAccount linkedAccount, OAuth2User userInfo) {
+    var passportJwtString = userInfo.<String>getAttribute(PASSPORT_JWT_V11_CLAIM);
+    if (passportJwtString != null) {
+      var passportJwt = decodeJwt(passportJwtString);
+
+      var visaJwtStrings =
+          Objects.requireNonNullElse(
+              passportJwt.getClaimAsStringList(GA4GH_PASSPORT_V1_CLAIM),
+              Collections.<String>emptyList());
+
+      var visas =
+          visaJwtStrings.stream().map(s -> buildVisa(decodeJwt(s))).collect(Collectors.toList());
+
+      return new LinkedAccountWithPassportAndVisas.Builder()
+          .linkedAccount(linkedAccount)
+          .passport(buildPassport(passportJwt))
+          .visas(visas)
+          .build();
+    } else {
+      return new LinkedAccountWithPassportAndVisas.Builder().linkedAccount(linkedAccount).build();
+    }
+  }
+
+  private static GA4GHPassport buildPassport(Jwt passportJwt) {
+    var passportExpiresAt = getJwtExpires(passportJwt);
+
+    return new GA4GHPassport.Builder()
+        .jwt(passportJwt.getTokenValue())
+        .expires(passportExpiresAt)
+        .build();
+  }
+
+  private static GA4GHVisa buildVisa(Jwt visaJwt) {
+    JSONObject visaClaims = getJwtClaim(visaJwt, GA4GH_VISA_V1_CLAIM);
+    var visaType = visaClaims.get(VISA_TYPE_CLAIM);
+    if (visaType == null) {
+      throw new InvalidJwtException(String.format("visa missing claim [%s]", VISA_TYPE_CLAIM));
+    }
+    return new Builder()
+        .visaType(visaType.toString())
+        .jwt(visaJwt.getTokenValue())
+        .expires(getJwtExpires(visaJwt))
+        .issuer(visaJwt.getIssuer().toString())
+        .lastValidated(new Timestamp(Instant.now().toEpochMilli()))
+        .tokenType(determineTokenType(visaJwt))
+        .build();
+  }
+
+  private static Timestamp getJwtExpires(Jwt decodedPassportJwt) {
+    var expiresAt = decodedPassportJwt.getExpiresAt();
+    if (expiresAt == null) {
+      throw new InvalidJwtException("jwt missing expires (exp) claim");
+    }
+    return new Timestamp(expiresAt.toEpochMilli());
+  }
+
+  private static <T> T getJwtClaim(Jwt jwt, String claimName) {
+    T claim = jwt.getClaim(claimName);
```

Comment:
Can do this in a single call
```suggestion
    return Optional.ofNullable(jwt.getClaim(claimName)).orElseThrow(() -> new InvalidJwtException(String.format("jwt missing claim [%s]", claimName)));
```